File: solutions/JavaScript/sum_root_to_leaf_numbers_dfs.js
/**
 * Definition for a binary tree node.
 */
class TreeNode {
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

/**
 * Sum all root-to-leaf numbers using DFS.
 * Build number by appending digits as we traverse down.
 * Time: O(n), Space: O(h) for recursion stack
 * @param {TreeNode} root
 * @return {number}
 */
function sumNumbers(root) {
    function dfs(node, currentSum) {
        if (!node) return 0;

        // Build number: multiply by 10 and add current digit
        currentSum = currentSum * 10 + node.val;

        // Leaf node: return the complete number
        if (!node.left && !node.right) {
            return currentSum;
        }

        // Recursively process children and sum
        return dfs(node.left, currentSum) + dfs(node.right, currentSum);
    }

    return dfs(root, 0);
}

// Test cases
const root1 = new TreeNode(1);
root1.left = new TreeNode(2);
root1.right = new TreeNode(3);

console.log("Example 1 sum:", sumNumbers(root1));  // 25 (12 + 13)

const root2 = new TreeNode(4);
root2.left = new TreeNode(9);
root2.right = new TreeNode(0);
root2.left.left = new TreeNode(5);
root2.left.right = new TreeNode(1);

console.log("Example 2 sum:", sumNumbers(root2));  // 1026 (495 + 491 + 40)

File: solutions/Rust/sum_root_to_leaf_numbers_dfs.rs
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

/**
 * Sum all root-to-leaf numbers using DFS.
 * Build number by appending digits as we traverse down.
 * Time: O(n), Space: O(h) for recursion stack
 */
pub fn sum_numbers(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    fn dfs(node: Option<Rc<RefCell<TreeNode>>>, current_sum: i32) -> i32 {
        match node {
            None => 0,
            Some(n) => {
                let node_ref = n.borrow();
                let val = node_ref.val;
                let left = node_ref.left.clone();
                let right = node_ref.right.clone();
                drop(node_ref);

                // Build number: multiply by 10 and add current digit
                let current_sum = current_sum * 10 + val;

                // Leaf node: return the complete number
                if left.is_none() && right.is_none() {
                    return current_sum;
                }

                // Recursively process children and sum
                dfs(left, current_sum) + dfs(right, current_sum)
            }
        }
    }

    dfs(root, 0)
}

fn main() {
    println!("Example 1: Sum root to leaf numbers with DFS");
}
